### PR TITLE
Add transaction when creating pull request to avoid creating dirty data

### LINKED ISSUE #26129: Git Hooks (pre-receive) with fail check drops Gitea to 502 Error
### Description

I have faced a situation where Git Hook drops Gitea to 502 Error.
Git Hook checks commits for some naming templates and raise error if check is not completed. All commits/PRs are blocked if check is failing.

Issue:
Gitea executes Git Hook on PR with non 0 error code. After receiving exit code 1 it drops to 502 Error

Expected behavior:
Show page with git console and error code from git. Do not drop to 502. 

How to reproduce:
1. Enable git hooks on Gitea and Configure Git Hook (pre-recive) on server side with this code:
```
#!/bin/bash

# Error message for BRANCH POLICY
error_msg_branch=$(cat <<-END

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@                                                                               @   
@   !!! Push not allowed by BRANCH NAME policy !!!                              @
@                                                                               @  
@   You branch should be named with these templaes:                             @
@     - master                                                                  @
@     - develop                                                                 @
@     - release/RC_*                                                            @
@     - feature/JIRA_ID-000--*                                                  @
@     - bugfix/*                                                                @
@     - hotfix/*                                                                @
@                                                                               @
@   Example: feature/MEC-167--add_new_functionaliry                             @
@                                                                               @
@   Wiki: https://confluence.ncloudtech.ru/display/KDM/Git+branch+naming        @                                                                        @
@                                                                               @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    
END
)

error_msg_commit=$(cat <<-END

@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@                                                                                 @
@   !!! Your Commit does not have a referance to JIRA ticket !!!                  @
@                                                                                 @
@   Please correct your Git messages, and push again.                             @
@   Example: "[KDM-1234] This is a correct JIRA reference"                        @
@                                                                                 @
@   Wiki: https://confluence.ncloudtech.ru/pages/viewpage.action?pageId=201708163 @                                                                                @
@                                                                                 @
@   To FORCE push, use "bugfix" or "hotfix" in your commit message                @                                                                  @
@                                                                                 @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
    
END
)



while read oldrev newrev refname
do
  #
  # Step 0:
  # Get Git commit information
  #
  BRANCH_NAME_FULL=$refname
  BRANCH_NAME="$(echo $BRANCH_NAME_FULL | sed 's/refs\/heads\///g')"
  COMMIT_MESSAGE=$(git log --format=%B -n 1)
  echo "[INFO] BRANCH NAME: $BRANCH_NAME"

  #
  # Step 1:
  # Policy - Check that brunch name is enforced by branch name
  #

  echo "[INFO] Policy - Check that brunch name is enforced by branch name"
  # Regexp for allowed names of branches
  branch_name_format='^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*'
  
  if [[ ! $BRANCH_NAME =~ $branch_name_format ]]; then
      echo "$error_msg_branch" >&2
      exit 1
  else
      echo "Push is successful"
  fi

  #
  # Step 2:
  # Policy - Check commit message for JIRA issue number
  #

  # Configuration
  echo "[INFO] Policy - Check commit message for JIRA issue number"
  jiraIdRegex="[a-zA-Z0-9,\.\_\-]+-[0-9]+"
  fixMsgRegex="bugfix|hotfix"
  info_msg="[INFO] The commit message looks good"
  error_msg="[POLICY] The commit doesn't reference a JIRA issue"

  # Get all commits from this push
  for sha1Commit in $(git rev-list $oldrev..$newrev);
  do
    # Receive git commits sha from git history in chronologic order
    echo "[INFO] Processing commit with sha: $sha1Commit";
    # Get commit message from commit
    commitMessage=$(git log --format=%B -n 1 $sha1Commit)
    # Check with RegEX if commit has JIRA reference
    jiraIds=$(echo $commitMessage | grep -Eo $jiraIdRegex)
    fixMsg=$(echo $commitMessage | grep -Eo $fixMsgRegex)
    # Check if this commit urgent e.g. hotfox or bugfix
    if [[ -n "${fixMsg}" ]]; then
      echo "[WARNING] Found "bugfix|hotfix" in msg. Force skipping check for JiraID"
      exit 0
    fi
    # Check for JiraIDs in commit message
    echo "[INFO] Found JIRA IDs in commit: $jiraIds"
    if [[ -z "${jiraIds}" ]]; then
      echo "$error_msg: $commitMessage" >&2
      echo "$error_msg_commit" >&2
      exit 1
    fi
  done
done

#
# Exit
#
exit 0
```
2. Create new branch `feature/jira-123_some_text` and push to gitea
3. Create PR from `feature/jira-123_some_text` --> master
4. Get Error 502 on Gitea. The reason of this error, bacause git hook check failes. It not allows refs with PRs and exit with error 1. 
I expect, that it will show `$error_msg_branch >&2` from bash script in hook redirected to git log (How it is done on client side, if hook failes), but not prop to 502 Error

If i change   `branch_name_format='^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*'` to `  branch_name_format='pull|^master|^main|^develop|^release\/RC_.*|^feature\/[a-zA-Z0-9,\.\_\-]+-[0-9]+.*|^hotfix\/.*|^bugfix\/.*' the check will not fail and PR working
`

### Gitea Version

1.20.0

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

Screenshot1
Screenshot2

### Git Version

_No response_

### Operating System

linux

### How are you running Gitea?

docker

### Database

PostgreSQL

### PR DESCRIPTION
**Pull request by lunny**
Fix #26129
Replace #26258 

This PR will introduce a transaction on creating pull request so that if some step failed, it will rollback totally. And there will be no dirty pull request exist.

### CODE REVIEW

**Review by wxiaoguang — COMMENTED** [review]
How does the new code address the comment?

```
	// Now - even if the request context has been cancelled as the PR has been created
	// in the db and there is no way to cancel that transaction we have to proceed - therefore
	// create new context and work from there
	prCtx, _, finished := process.GetManager().AddContext(graceful.GetManager().HammerContext(), fmt.Sprintf("NewPullRequest: %s:%d", repo.FullName(), pr.Index))
	defer finished()
```

**Comment by lunny** [comment]
> How does the new code address the comment?
> 
> ```
> 	// Now - even if the request context has been cancelled as the PR has been created
> 	// in the db and there is no way to cancel that transaction we have to proceed - therefore
> 	// create new context and work from there
> 	prCtx, _, finished := process.GetManager().AddContext(graceful.GetManager().HammerContext(), fmt.Sprintf("NewPullRequest: %s:%d", repo.FullName(), pr.Index))
> 	defer finished()
> ```

prCtx has been used on the database trasaction function with opening a repository, if it's failed, the database transaction will rollback.

**Comment by wxiaoguang** [comment]
> prCtx has been used on the database trasaction function with opening a repository, if it's failed, the database transaction will rollback.

Why `prCtx` is necessary? Is there anything wrong to use `ctx` to replace the `prCtx`?

**Comment by lunny** [comment]
> > prCtx has been used on the database trasaction function with opening a repository, if it's failed, the database transaction will rollback.
> 
> Why `prCtx` is necessary? Is there anything wrong to use `ctx` to replace the `prCtx`?

`process.GetManager().AddContext` will only be used to track the pull request creation now.

File: services/pull/pull.go [hunk 1]
```diff
@@ -38,7 +38,11 @@ import (
 var pullWorkingPool = sync.NewExclusivePool()
 
 // NewPullRequest creates new pull request with labels for repository.
-func NewPullRequest(ctx context.Context, repo *repo_model.Repository, pull *issues_model.Issue, labelIDs []int64, uuids []string, pr *issues_model.PullRequest, assigneeIDs []int64) error {
+func NewPullRequest(ctx context.Context, repo *repo_model.Repository, issue *issues_model.Issue, labelIDs []int64, uuids []string, pr *issues_model.PullRequest, assigneeIDs []int64) error {
+	// add the creation of pull request to process management
+	ctx, _, finished := process.GetManager().AddContext(ctx, fmt.Sprintf("NewPullRequest: %s:%d", repo.FullName(), pr.Index))
+	defer finished()
```
Comment (wxiaoguang) [inline services/pull/pull.go]: I would suggest to remove this `AddContext`, because the parent context (request context) has already been managed by the "process manager" and it already have its name and labels.

Adding more context seems redundancy

File: services/pull/pull.go [hunk 2]
```diff
@@ -118,21 +110,52 @@ func NewPullRequest(ctx context.Context, repo *repo_model.Repository, pull *issu
...
 		}
+		return nil
+	}); err != nil {
+		// cleanup: this will only remove the reference, the real commit will be clean up when next GC
+		if err1 := baseGitRepo.RemoveReference(pr.GetGitRefName()); err1 != nil {
+			log.Error("RemoveReference: %v", err1)
+		}
+		return err
+	}
+	baseGitRepo.Close()
```
Comment (wxiaoguang) [inline services/pull/pull.go]: This line seems incorrect? Delete it?

File: services/pull/pull.go [hunk 3]
```diff
@@ -50,61 +54,49 @@ func NewPullRequest(ctx context.Context, repo *repo_model.Repository, pull *issu
...
-	mentions, err := issues_model.FindAndUpdateIssueMentions(ctx, pull, pull.Poster, pull.Content)
-	if err != nil {
-		return err
-	}
+		for _, assigneeID := range assigneeIDs {
+			comment, err := issue_service.AddAssigneeIfNotAssigned(ctx, issue, issue.Poster, assigneeID, false)
+			if err != nil {
+				return err
+			}
+			assigneeCommentMap[assigneeID] = comment
```
Comment (wxiaoguang) [inline services/pull/pull.go:80]: According to the `AddAssigneeIfNotAssigned` behavior, the "comment" and "err" could be both `nil`.

File: services/pull/pull.go (hunk 2 above)
Comment (lunny) [reply services/pull/pull.go]: Added some comments.

File: services/pull/pull.go (hunk 1 above)
Comment (lunny) [reply services/pull/pull.go]: Fixed

File: services/pull/pull.go (hunk 3 above)
Comment (lunny) [reply services/pull/pull.go:80]: yes, it maybe nil. It has been checked by notifications functions

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM （the comment for `baseGitRepo.Close()` is not clear to me)

File: services/pull/pull.go [hunk 4]
```diff
@@ -38,73 +37,70 @@ import (
...
 
 // NewPullRequest creates new pull request with labels for repository.
-func NewPullRequest(ctx context.Context, repo *repo_model.Repository, pull *issues_model.Issue, labelIDs []int64, uuids []string, pr *issues_model.PullRequest, assigneeIDs []int64) error {
-	if err := TestPatch(pr); err != nil {
-		return err
-	}
-
-	divergence, err := GetDiverging(ctx, pr)
+func NewPullRequest(ctx context.Context, repo *repo_model.Repository, issue *issues_model.Issue, labelIDs []int64, uuids []string, pr *issues_model.PullRequest, assigneeIDs []int64) error {
+	prCtx, cancel, err := createTemporaryRepoForPR(ctx, pr)
```
Comment (wxiaoguang) [inline services/pull/pull.go:41]: No idea why this change is a must.

The `TestPatch` could be reused by adding a `ctx`

Comment (lunny) [reply services/pull/pull.go:41]: > No idea why this change is a must.
> 
> The `TestPatch` could be reused by adding a `ctx`

Because both `TestPatch` and `GetDiverging` will invoke `createTemporaryRepoForPR` which should be create only once.

**Review by wolfogre — APPROVED** [review]
